src/System.Reflection.Metadata/src/System/Reflection/Metadata/Handles.cs

```diff
@@ -96,30 +198,165 @@ internal static void ThrowInvalidCodedIndex()
+    {
+        // bits:
+        //     31: IsVirtual
+        // 24..30: type
+        //  0..23: row id
+        private readonly uint _vToken;
+
+        internal EntityHandle(uint vToken)
+        {
+            _vToken = vToken;
+        }
+
+        public static implicit operator Handle(EntityHandle handle)
+        {
+            return Handle.FromVToken(handle._vToken);
+        }
+
+        public static explicit operator EntityHandle(Handle handle)
+        {
+            if (handle.IsHeapHandle)
+            {
+                Handle.ThrowInvalidCast();
+            }
+
+            return new EntityHandle(handle.EntityHandleValue);
+        }
+
+        internal uint Type
+        {
+            get { return _vToken & TokenTypeIds.TypeMask; }
+        }
+
+        internal uint VType
+        {
+            get { return _vToken & (TokenTypeIds.VirtualBit | TokenTypeIds.TypeMask); }
+        }
+
+        internal bool IsVirtual
+        {
+            get { return (_vToken & TokenTypeIds.VirtualBit) != 0; }
+        }
+
+        public bool IsNil
+        {
+            // virtual handle is never nil
+            get { return (_vToken & (TokenTypeIds.VirtualBit | TokenTypeIds.RIDMask)) == 0; }
+        }
+
+        internal int RowId
+        {
+            get { return (int)(_vToken & TokenTypeIds.RIDMask); }
+        }
+
+        /// <summary>
+        /// Value stored in a specific entity handle (see <see cref="TypeDefinitionHandle"/>, <see cref="MethodDefinitionHandle"/>, etc.).
+        /// </summary>
+        internal uint SpecificHandleValue
+        {
+            get { return _vToken & (TokenTypeIds.VirtualBit | TokenTypeIds.RIDMask); }
+        }
+
+        public HandleKind Kind
+        {
+            get { return HandleType.ToHandleKind(Type >> TokenTypeIds.RowIdBitCount); }
+        }
+
+        internal int Token
+        {
+            get
+            {
+                Debug.Assert(!IsVirtual);
+                return (int)_vToken;
+            }
+        }
+
+        public override bool Equals(object obj)
+        {
+            return obj is EntityHandle && Equals((EntityHandle)obj);
+        }
+
+        public bool Equals(EntityHandle other)
+        {
+            return _vToken == other._vToken;
+        }
+
+        public override int GetHashCode()
+        {
+            return unchecked((int)_vToken);
+        }
+
+        public static bool operator ==(EntityHandle left, EntityHandle right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(EntityHandle left, EntityHandle right)
+        {
+            return !left.Equals(right);
+        }
+
+        internal static int Compare(EntityHandle left, EntityHandle right)
+        {
+            // all virtual tokens will be sorted after non-virtual tokens
+            return (int)((left._vToken & TokenTypeIds.RIDMask) | ((left._vToken & TokenTypeIds.VirtualBit) >> 3)) -
```

Comment:
Pre-existing, I believe, but shouldn't the entity type factor in to these comparisons?